### DISCUSSION TITLE
prevent stealth theft of mountpoints.  I think its what is happening

### DISCUSSION THREAD

**Discussion by hput** [psy0rz/zfs_autobackup#132] [Q&A]
I seem to be having more trouble than the average bear getting on to how to use zfs_autobackup.
My test setup has once again become unusable.  It consists of 2 VBox vms and two real hardware  hosts that act as the --ssh-source hosts
So briefly I followed the steps laid out in the wiki
first marked the zfs fs to be backedup (replicated I guess its called)
In the wiki its like this:
       zfs set autobackup:offsite1=true rpool
In my test example it was a little more ambitious and named differently
        I seem to be having more trouble than the average bear getting on to how to use zfs_autobackup.
My test setup has once again become unusable.  It consists of 2 VBox vms and two real hardware  hosts that act as the --ssh-source hosts
So briefly I followed the steps laid out in the wiki
first marked the zfs fs to be backedup (replicated I guess its called)
In the wiki its like this:
       zfs set autobackup:offsite1=true rpool
In my test example it was a little more ambitious and named differently
       So on the ssh-source host
       zfs set autobackup:snr=true rpool p0 p1 bpool
     Followed by the step where you pull  backup
From wiki:
zfs-autobackup -v --ssh-source pve01 offsite1 data/backup/pve01
My test setup: The command below is being run from VBox vm in role of backup server.

   (OS on backup server is openindiana) [An solaris like offshoot from OpenSolaris] 

   zfs-autobackup -v --ssh-source 2x snr p2/42/za/2x

This all went off well and collected a thorough starter backup.

The trouble starts when 2 days later I get around to rebooting backup server

The reboot hits some snags and offers  login for root to have a shot at it.
The major complaint seems to be:  
   "Mounting zfs filesystems: svc:/system/filesysystem/local:default: WARNING: /usr/sbin/zfs mount -a failed:
Exit status 1
![VirtualBox_oi2_07_05_2022_20_14_48-mount-failed](https://user-images.githubusercontent.com/104574647/167270842-46616618-2e66-4df0-919d-2dd6ed6962c7.png)

You can see in the upper part a list of `cannot mount' Directory is not empty.  All the mountpoints are already taken somehow. The replicated zfs fs.  How do you forestall something like this happening.  I don't recall seeing anything in the wiki docs about changing mountpoints, and of course all the replicated fs do have mount points. 
I can still login to the VM so please let me know what to run to gather information you can use

I've generated a display with zfs list -r 
[zfslist-r.txt](https://github.com/psy0rz/zfs_autobackup/files/8645914/zfslist-r.txt)

Everything under pool "p2" is a replicated FS and you can see by the mountpoints that the replicated stuff is setting where OS filesystems should be.

I can get it all back I think but need to know where I went so badly wrong

**Comment by psy0rz** [comment]
like the tips say:

> Use --clear-mountpoint to prevent the target machine from mounting the backupped filesystem in the wrong place during a reboot.

Just set the canmount option to noauto for those datasets and reboot. Then they wont mount and it will be fine.

**Reply by hput** [reply]
I ran up on that `--clear-mountpoint' a while after posting but it is not explained at all in the tip   Where it gits used how to use it, some example like with all the other steps seems in order. 
What happens when an zfs FS  has no mountpoint or a cleared mountpoint.  Does that not render the FS sort of useless.
Would you just leave it unmounted from then on?

**Reply by hput** [reply]
Sorry about coming off as genuine pita.  I've never been known for great native intelligence.
I'm not getting what to do with all those canmount=noauto fs.  Since they are replication zfs fs I would not want to mess with them unless for restore  Or maybe something less than full restore where a few or even 1 specific file needs to be copied into a host somewhere on network  
The full meaning of having the replicated fs on hand is only now beginning to take on some meaning to me.
I'm trying out at the same time using zfs-backup for general snapshotting.
I noticed in output during running snapshots something about there being no changes since previous snap and thus being bypassed. Am I right thinking that zfs-autobackup when used as snapshottting agent only, creates a snap only if there has been change since the last one? 

Having been a sanoid user for a number of months I'm pleased to see that conserving work and space is something zfs-autobackup  does by default.

It would be most welcome if you can take the time to verify my thinking, or once again pull my head from where the sun don't shine.




.

**Comment by Scrin** [comment]
It's more about "general ZFS knowledge" (mostly regarding transferring datasets between systems) and the right choice of options greatly depends on the setup and what you want to achieve.

That being said, if you have explicit mountpoints (=not inherited from parent dataset) on the datasets to back up, you might want to clear the mountpoints (with the `--clear-mountpoint` option) if you have automount for ZFS on the target host to avoid mounting the backups which can lead to mistakes since you could accidentally tamper with the backups.

On the other hand if you want to preserve the mountpoint property with the backed up dataset, you should probably use other means to prevent automatically mounting the datasets on the target host, and the ideal way to do this again depends on your general setup. For example, if the target host does not use ZFS for anything but holding the backups, disabling zfs automount globally on boot could be an option. Another option could be using the [canmount](https://docs.oracle.com/cd/E19253-01/819-5461/gdrcf/index.html) property, which can be set with the `--set-properties` option (although I haven't tested this myself but it should work).

Another example is the kind of setup that I have; none of the datasets to back up have explicit mountpoints, they all inherit the mountpoint from their respective parent dataset, and on the backupper target host the parent dataset has no mountpoint so the backed up datasets inherit the "no mountpoint" and thus I don't need to clear the mountpoint or set the canmount property.

**Reply by hput** [reply]
Thank you for the thoughtful reply, I'm thinking the last situation you describe seems like the one I should shoot for.

Scrin Wrote:
> Another example is the kind of setup that I have; none of the datasets to back up have explicit mountpoints, they all inherit the mountpoint from their respective parent dataset, and on the backupper target host the parent dataset has no mountpoint so the backed up datasets inherit the "no mountpoint" and thus I don't need to clear the mountpoint or set the canmount property.

Am I wrong in thinking: "  None of the datasets to back up have explicit mountpoint ...."  Is somewhat unusual?  That is, wouldn't the average zfs host have explicit mountpoints as a matter of course?  Or is it the case that you administer the hosts to be backed up and the state they are configured at is part of your plan to help the replication along.
If so, I might be getting a glimmer of a broader plan that fits together nicely.

Lets say I start trying to use the setup you have described.  The zfs servers I have control of are strictly a local afair.
The zfs hosts do get data in and out as a matter of course... so it seems that would need actual mountpoints.

I'm not really sure of that.  Way to green to have hardly anything more than suspicions on how to use and manage zfs servers.  Really just tinkering and slowly finding my way.

I'm sorry that puts an inordinate load on folks like you and psy0rz. I don't really mean to be so tedious and dense.
I do thank you both for taking time to answer my questions and especially taking time for thoughtful answers.

**Reply by Scrin** [reply]
In cases where ZFS is used as the root filesystem (=the filesystem that is mounted at `/`), then yes, not having explicit mountpoints for at least some of the datasets is a somewhat unusual (or even impossible) scenario. However even in such scenario, not **all** datasets require an explicit mountpoint, and it again depends on your setup if it's a good thing or a bad thing to use explicit mountpoints rather than inherited ones in all or some of the datasets.

In my setup, the root filesystem of the servers I run is ephemeral; I don't care about them, I can lose and recreate them at any time without losing any important data, and the only data I care about is under a specific path (or a few). The most simple example about this is how my "generic docker hosts" are set up; they only run regular "simple" docker containers which store their data under a specific path like this:

All container data volumes are stored under `/docker/data/<name_of_the_service>/<name_of_the_volume>`
All container (generated) configuration is under `/docker/config/<name_of_the_service>`
And so on..

In this setup I have a "docker data root dataset" with an **explicit** mountpoint at `/docker`, and below this there are child datasets such as `data` and `config` that have no explicit mountpoint set; instead they inherit their mountpoint from the parent dataset and so on (ie. the paths mentioned above), and since the only "critical" data resides somewhere under `/docker/data/` path, that's the only dataset I back up, recursively (each service has its own dataset under this "data" dataset, in order to set service-specific quotas and track disk usage per service etc, and most importantly so that I can rollback a specific service without affecting the others), so in this case none of the datasets I back up have an explicit mountpoint, since I don't backup the "docker data root dataset" itself (the one explicitly mounted at `/docker` in my case).

In the above setup I find it a good thing not to have any other explicit mountpoints apart from the "docker root dataset"; this way I can change the mountpoint of the entire tree "at once" by just changing the mountpoint of the root dataset. Although the need of changing this mountpoint is very unlikely, I don't see any benefit in using explicit mountpoints for the child datasets (in my setup) either since it'd be just unnecessary extra config to take care of, since by default datasets inherit their mountpoint from their parent dataset mountpoint, and appending their name to that.

For example, a dataset at `foo/bar/baz` will by default inherit it's mountpoint by taking the mountpoint of `foo/bar` dataset and appending `/baz` to that; ie. if `foo/bar` is mounted at `/some/path/`, then `foo/bar/baz` would inherit its mountpoint to be `/some/path/baz/`, which is exactly the directory structure that I want anyway.
